packages/dev/core/src/Rendering/fluidRenderer/fluidRenderer.ts

```diff
@@ -0,0 +1,547 @@
+
+        this._renderObjects.splice(index, 1);
+
+        if (removeUnusedTargetRenderer && this._removeUnusedTargetRenderers()) {
+            this._initialize();
+        } else {
+            this._setParticleSizeForRenderTargets();
+        }
+
+        return true;
+    }
+
+    private _sortRenderingObjects(): void {
+        this._renderObjects.sort((a, b) => {
+            return a.object.priority < b.object.priority ? -1 : a.object.priority > b.object.priority ? 1 : 0;
+        });
+    }
+
+    private _removeUnusedTargetRenderers(): boolean {
+        const indexes: { [id: number]: boolean } = {};
+
+        for (let i = 0; i < this._renderObjects.length; ++i) {
+            const targetRenderer = this._renderObjects[i].targetRenderer;
+            indexes[this._targetRenderers.indexOf(targetRenderer)] = true;
+        }
+
+        let removed = false;
+        const newList: Array<FluidRenderingTargetRenderer> = [];
+        for (let i = 0; i < this._targetRenderers.length; ++i) {
+            if (!indexes[i]) {
+                this._targetRenderers[i].dispose();
+                removed = true;
+            } else {
+                newList.push(this._targetRenderers[i]);
+            }
+        }
+
+        if (removed) {
+            this._targetRenderers.length = 0;
+            this._targetRenderers.push(...newList);
+        }
+
+        return removed;
+    }
+
+    private static _IsParticleSystemObject(obj: FluidRenderingObject): obj is FluidRenderingObjectParticleSystem {
+        return !!(obj as FluidRenderingObjectParticleSystem).particleSystem;
+    }
+
+    private _getParticleSystemIndex(ps: IParticleSystem): number {
+        for (let i = 0; i < this._renderObjects.length; ++i) {
+            const obj = this._renderObjects[i].object;
+            if (FluidRenderer._IsParticleSystemObject(obj) && obj.particleSystem === ps) {
+                return i;
+            }
+        }
+
+        return -1;
+    }
+
+    private _initialize(): void {
+        for (let i = 0; i < this._targetRenderers.length; ++i) {
+            this._targetRenderers[i].dispose();
+        }
+
+        const cameras: Map<Camera, [Array<FluidRenderingTargetRenderer>, { [key: string]: FluidRenderingDepthTextureCopy }]> = new Map();
+
+        for (let i = 0; i < this._targetRenderers.length; ++i) {
+            const targetRenderer = this._targetRenderers[i];
+
+            targetRenderer._initialize();
+
+            if (targetRenderer.camera && targetRenderer._renderPostProcess) {
+                let list = cameras.get(targetRenderer.camera);
+                if (!list) {
+                    list = [[], {}];
+                    cameras.set(targetRenderer.camera, list);
+                }
+                list[0].push(targetRenderer);
+                targetRenderer.camera.attachPostProcess(targetRenderer._renderPostProcess, i);
+            }
+        }
+
+        cameras.forEach((list, camera) => {

```

Comment:
See the answer above, I can't use the `for .. of` construct. The alternative is to use an iterator "by hand", which I don't like because I find it not very readable...

This path is not a hot path, which means that it is not called on every frame, only when things need to be reset, which rarely (if ever) happens, when you update a property like the width/height of a texture for example.

Let me know if it's ok for you.